lib/spack/spack/cmd/__init__.py

```diff
@@ -46,29 +47,31 @@ def cmd_name(python_name):
 
 #: global, cached list of all commands -- access through all_commands()
 _all_commands = None
+_default_command_path = spack.paths.command_path
 
 
-def all_commands():
-    """Get a sorted list of all spack commands.
+def all_commands(command_list, command_path=None):
```

Comment:
The previous iteration of `all_commands()` unconditionally cached the result in `_all_commands`, returning it unaltered if already set. The new version receives the cache variable, filling it if empty and returning it unaltered if not. Since the new version is flexible with respect to the command path searched, it needs to be flexible with respect to a caller-provided cache variable to avoid returning the wrong answer.

Reply:
Note that the new `spack.main.all_commands()` provides the old behavior of `spack.cmd.all_commands()` by providing its own `_all_commands` cache variable.